### PR TITLE
ceph-volume: do not convert a LV's symlink to real path

### PR DESCRIPTION
**Pull request by yician1000**
The only valid LV path format to the filter key 'lv_path' is "/dev/vg/lv" but not 
"/dev/mapper/vg-lv". Therefore, do not convert a symlink to its real path when 
the symlink is a LV.

Fixes: https://tracker.ceph.com/issues/61597

Signed-off-by: Jerry Pu <jerrypu@qnap.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by yician1000** [comment]
The fix is to check if the give path is a LV first. If the path represents a LV, then do not convert the path to its real path.

**Comment by yician1000** [comment]
jenkins test make check

**Comment by guits** [comment]
Thanks for your patch @yician1000
That looks good. We should probably add a unit test for this

**Comment by yician1000** [comment]
@guits OK, I will help to write the unit test.

**Comment by guits** [comment]
jenkins test ceph-volume tox

**Comment by yician1000** [comment]
jenkins test ceph-volume all

**Comment by yician1000** [comment]
jenkins test ceph-volume tox

**Comment by yician1000** [comment]
jenkins test ceph-volume tox

**Comment by yician1000** [comment]
jenkins retest this please

**Comment by yician1000** [comment]
jenkins retest this please

**Comment by yician1000** [comment]
jenkins retest this please

**Comment by yician1000** [comment]
jenkins test ceph-volume tox

**Comment by yician1000** [comment]
jenkins test ceph-volume tox

**Comment by yician1000** [comment]
Hi @guits 

I've added a unit test into this patch. Could you help to review it? Thanks.

**Comment by yician1000** [comment]
Hi @guits 

Could you help to check the patch? Any feedback is appreciated. Thanks!

**Comment by yician1000** [comment]
ping

**Comment by Svelar** [comment]
jenkins test ceph-volume all

**Comment by Svelar** [comment]
jenkins test ceph-volume tox

**Comment by yician1000** [comment]
jenkins test ceph-volume tox

**Comment by guits** [comment]
https://pulpito.ceph.com/gabrioux-2024-07-01_14:07:31-orch:cephadm-wip-guits-main-2024-07-01-0926-distro-default-smithi/

**Comment by yician1000** [comment]
Hi @guits 
I am busy on my other work and have no free time to handle the issue recently. Thank you for helping handle the issue.
 
It seems that there is some problem with the patch? I can see an error message in the Pulpito dashboard but I have no idea why it is.

> cluster [WRN] Health check failed: failed to probe daemons or devices (CEPHADM_REFRESH_FAILED)

**Comment by guits** [comment]
@yician1000 do you still hit this issue with ceph@main ?

**Comment by guits** [comment]
I made some tests this morning, the initial issue you reported seems already addressed.
That being said, there are still other issues when setting `DM_DISABLE_UDEV`:

for example:
```
# ceph-volume lvm zap --destroy --osd-id 0
--> Zapping: /dev/mapper/ceph--b647db3f--dba9--42ea--93c1--7e124d0a0b3a-osd--db--f083f526--268e--43af--8033--01ee31521104
--> Refusing to zap the mapper device: <Unknown: /dev/mapper/ceph--b647db3f--dba9--42ea--93c1--7e124d0a0b3a-osd--db--f083f526--268e--43af--8033--01ee31521104>
```

File: src/ceph-volume/ceph_volume/util/device.py [hunk 1]
```diff
@@ -108,8 +108,9 @@ def __init__(self, path, with_lsm=False, lvs=None, lsblk_all=None, all_devices_v
         if os.path.islink(self.path):
             self.symlink = self.path
             real_path = os.path.realpath(self.path)
-            # check if we are not a device mapper
-            if "dm-" not in real_path:
+            lv = lvm.get_single_lv(filters={'lv_path': self.path})
```
Comment (guits) [inline src/ceph-volume/ceph_volume/util/device.py:111]: this implies a new subprocess call in `Device()` that I'd like to avoid

**Comment by guits** [comment]
@yician1000 FYI, I reworked the patch here #58424

**Comment by yician1000** [comment]
Hi @guits 
> @yician1000 do you still hit this issue with ceph@main ?

I didn't try that with latest ceph@main.

> @yician1000 FYI, I reworked the patch here https://github.com/ceph/ceph/pull/58424

Thank you for your effort!

File: src/ceph-volume/ceph_volume/util/device.py (hunk 1 above)
Comment (yician1000) [reply src/ceph-volume/ceph_volume/util/device.py:111]: Got it. That makes sense.

**Comment by yician1000** [comment]
@guits helps to rework the patch at https://github.com/ceph/ceph/pull/58424, so we no longer need the PR.

**Pull request closed without merging**